AirCasting/Services/OldMeasurementRemovers/DefaultRemoveOldMeasurementsService.swift

```diff
@@ -0,0 +1,40 @@
+import Foundation
+import CoreData
+
+protocol RemoveOldMeasurements {
+    func removeOldestMeasurements(in context: NSManagedObjectContext, from sessionUUID: SessionUUID) throws
+}
+
+final class DefaultRemoveOldMeasurementsService: RemoveOldMeasurements {
+    
+    enum Error: Swift.Error {
+        case removingMeasurementsSurplusFailed
+        case noStreamInSession
+    }
+    
+    /// In fixed and external sessions we need to remove measurements older than 24h, and we treat 24 like a date.
+    /// We know the time of the last measurement and based on that we are subtracking 24h in seconds.
+    func removeOldestMeasurements(in context: NSManagedObjectContext, from sessionUUID: SessionUUID) throws {
+        if let session = try? context.existingSession(uuid: sessionUUID) {
+            try removeMeasurementsSurplus(context: context, streams: session.allStreams)
+        } else {
+            let session = try context.existingExternalSession(uuid: sessionUUID)
+            try removeMeasurementsSurplus(context: context, streams: session.allStreams)
+        }
+    }
+    
+    private func removeMeasurementsSurplus(context: NSManagedObjectContext, streams: [MeasurementStreamEntity]) throws {
+        for stream in streams {
+            timeBasedRemover(context: context, stream: stream)
+        }
+    }
+    
+    private func timeBasedRemover(context: NSManagedObjectContext, stream: MeasurementStreamEntity) {
+        guard let lastMeasurement = stream.allMeasurements?.last else { Log.error("No last measurement when trying to remove those from > 24h"); return }
+        let twentyFour = 86400000 // 24 hours in miliseconds: 60 * 60 * 24
+        let beginingOfCorrectPeriod = lastMeasurement.time.milliseconds - twentyFour
+        stream.allMeasurements?.forEach({ measurement in
```

Comment:
I meant that instead of going through all of the 1400 measurements and checking the time we could let core data handle it with a fetch request like this:

```
private func timeBasedRemover(context: NSManagedObjectContext, stream: MeasurementStreamEntity) {
        guard let lastMeasurementTime = stream.allMeasurements?.last?.time else { Log.error("No last measurement when trying to remove those from > 24h"); return }
        let threshold = lastMeasurementTime.twentyFourHoursBefore
        
        let req: NSFetchRequest<MeasurementEntity> = NSFetchRequest(entityName: "MeasurementEntity")
        req.predicate = NSPredicate(format: "time < %@ AND measurementStream == %@", threshold as CVarArg, stream)
        if let measurements = try? context.fetch(req) {
            measurements.forEach { Log.info("Removing measurement: \($0)"); context.delete($0) }
        }
    }
```

(I added `twentyFourHoursBefore` to a Date extension)
I'm not an expert on performance in iOS but I think this could be faster.